app/components/privileged-action.js

```diff
@@ -0,0 +1,55 @@
+import { inject as service } from '@ember/service';
+import Component from '@glimmer/component';
+
+/**
+ * A component that wraps elements (probably mostly buttons in practice) that
+ * can be used to perform potentially privileged actions.
+ *
+ * This component requires a `userAuthorised` property, which must be a
+ * `boolean` indicating whether the user is authorised for this action. Admin
+ * rights need not be taken into account.
+ *
+ * If the current user is an admin and they have enabled sudo mode, then they
+ * are always allowed to perform the action, regardless of the return value of
+ * `userAuthorised`.
+ *
+ * There are three content blocks supported by this component:
+ *
+ * - `default`: required; this is displayed when the user is authorised to
+ *              perform the action.
+ * - `placeholder`: this is displayed when the user _could_ be authorised to
+ *                  perform the action (that is, they're an admin but have not
+ *                  enabled sudo mode), but currently cannot perform the action.
+ *                  If omitted, the `default` block is used with all form
+ *                  controls disabled and a tooltip added.
+ * - `unprivileged`: this is displayed when the user is not able to perform this
+ *                   action, and cannot be authorised to do so. If omitted, an
+ *                   empty block will be used.
+ *
+ * Note that all blocks will be output with a wrapping `<div>` for technical
+ * reasons, so be sure to style accordingly if necessary.
+ */
+export default class PrivilegedAction extends Component {
+  /** @type {import("../services/session").default} */
+  @service session;
+
+  /** @return {boolean} */
+  get isPrivileged() {
+    return this.session.isSudoEnabled || this.args.userAuthorised;
+  }
+
+  /** @return {boolean} */
+  get canBePrivileged() {
+    return !this.args.userAuthorised && this.session.currentUser?.is_admin && !this.session.isSudoEnabled;
+  }
+
+  /**
+   * @param {Element} element
+   * @return {void}
+   */
+  disableComponents(element) {
+    [...element.querySelectorAll('input, select, textarea, button')].forEach(control => {
+      control.setAttribute('disabled', 'disabled');
+    });
```

Comment:
I'm a bit worried about this part since it breaks the declarative nature of the template system. Did you test this in combination with a button that has `disabled` dynamically specified in the template too? It would make it a bit hard to understand what definition is "winning".

Reply:
If I understand how render modifiers work correctly (I probably don't), then `did-insert` fires after the Ember templating has executed, and therefore (at most) we'd be re-setting `disabled` on the element.

Empirically, hacking together a custom component with an element with a dynamic `disabled` attribute in the template and putting it inside a `PrivilegedAction` seems to work how I'd expect: if the placeholder state is active, then the element is disabled regardless of its own attribute.

I can't figure out a good way to express this in a unit test, but I'm open to suggestions.

Reply:
I did a little bit more research and it looks like another alternative could be https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset#disabled. It should allow us to replace the imperative code above with declarative template code.